### DISCUSSION TITLE
[apex] Can't get the value of N or M in the error message when violating Design-CyclomaticComplexity rule

### DISCUSSION THREAD

**Discussion by etingsvall20** [pmd/pmd#5157] [Q&A] [answered]
[apex]
 
**Affects PMD Version:** 7.0.0.
 
**Description:**
 
When executing the pmd on an apex class that violates the rule(threshold of >10) it displays this error message: [{
	"owner": "apex-pmd",
	"code": {
		"value": "CognitiveComplexity",
		"target": {
			"$mid": 1,
			"external": "[https://pmd.github.io/latest/pmd_rules_apex_design.html#cognitivecomplexity"](https://pmd.github.io/latest/pmd_rules_apex_design.html#cognitivecomplexity%22),
			"path": "/latest/pmd_rules_apex_design.html",
			"scheme": "https",
			"authority": "pmd.github.io",
			"fragment": "cognitivecomplexity"
		}
	},
	"severity": 4,
	"message": *"P3 - The method X has a cognitive complexity of N, current threshold is M (rule: Design-CognitiveComplexity)"*,
	"source": "apex pmd",
	"startLineNumber": 11,
	"startColumn": 5,
	"endLineNumber": 11,
	"endColumn": 42
}]

The class is a trigger handler with some maps with +10 fors+ifs in it, triggering the rule.

Is there a way to customize our ruleset in order to get the values instead of N and M? So we can analyze easily its performance.
 
Thank you.

**Comment by jsotuyod** [comment, marked as answer]
I'm not sure what you mean by "customize our ruleset in order to get the values instead of N and M"… the values are already part of the reported message.

You can customize that message, by overriding the `message` property in the rule reference, and using placeholders (just like the rule does):

https://github.com/pmd/pmd/blob/10add3f0f786a78ef6fff75717ba0d0c5c40a025/pmd-apex/src/main/resources/category/apex/design.xml#L162

but it will still be part of the message.

**Comment by etingsvall20** [comment]
Thank you very much!

I adapted it for both CyclomaticComplexity and CognitiveComplexity rules and works perfectly.
